[aws-lambda] add individual event types for all cognito triggers

netroy opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

darbio commented on types/aws-lambda/trigger/cognito-user-pool-trigger.d.ts:
```diff
@@ -1,102 +0,0 @@
-
-// Result type is weird: docs and samples say to return the mutated event, but it only requires an object
-// with a "response" field, the type of which is specific to the event.triggerType. Leave as any for now.
-export type CognitoUserPoolTriggerHandler = Handler<CognitoUserPoolTriggerEvent>;
```
Where is this reimplemented?

darbio commented on types/aws-lambda/trigger/cognito-user-pool-trigger.d.ts:
```diff
@@ -1,102 +0,0 @@
- * Cognito User Pool event
- * http://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html
- */
-export interface CognitoUserPoolTriggerEvent {
```
Where is this reimplemented?

darbio commented on types/aws-lambda/trigger/cognito-user-pool-trigger.d.ts:
```diff
@@ -1,102 +0,0 @@
-        };
-    };
-}
-export type CognitoUserPoolEvent = CognitoUserPoolTriggerEvent;
```
Where is this reimplemented?

netroy commented:
@darbio all the types you pointed out, are here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46142/files#diff-db38c6909822299c3bb14292b43f0f09R13
They are also backward compatible (I tested with the original tests). 
I have marked them as deprecated only to let the users know that there are better types available, if they wish to switch.

netroy commented on types/aws-lambda/trigger/cognito-user-pool-trigger/index.d.ts:
```diff
@@ -0,0 +1,49 @@
+ * @deprecated Please use specific event types instead
+ * @see http://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-identity-pools-working-with-aws-lambda-triggers.html
+ */
+export type CognitoUserPoolTriggerEvent =
```
This is in intersection type because the current implementation is equivalent of this intersection type. 
Ideally this should be a union type. but, that'd break the backward-compatibility.

simonbuchan commented:
I haven't had a chance to specifically review this, but I'm not sure that the "union" types should be deprecated: it's more efficient to implement closely related functions with one lambda that inspects the event as it is already warmed up.

I'm ok with it if there is a more type safe implementation now, I don't know if you can just intersect the specific handlers though?

netroy commented:
@simonbuchan I think what you are describing is exactly how I use these definitions. I have lambda functions that handle multiple types in a single functions. with union types, the moment I assert the value of `triggerSource`, the type collapses to a specific event type.

> I'm not sure that the "union" types should be deprecated
I haven't marked any union types as deprecated. 
It's the intersection types that are deprecated, where an event is of all the types at all times.

> I don't know if you can just intersect the specific handlers though?

This is possible now. All the handlers are exported, so people can combine them into their own custom union types.

simonbuchan commented:
Sounds great!

netroy commented:
if I remove the `tslint.json` changes, would that make merging this easier?

darbio commented:
You could try and remove/in-line the rules but I don't know if that will remove the need for a DT maintainer to do the merge seeing as the commits still exist in this PR which edit the file.

Give it a go and we'll see...

netroy commented:
@darbio I removed the `tslint` changes. I also removed the changes to the original types, and have simply marked them as `@deprecated`. This way we can be 100% sure that this won't break anything.

darbio commented:
I've reviewed. Let's see if it goes in.

darbio commented:
@netroy I'll let you do the honours 👍🏼🙌🏼

netroy commented:
Ready to merge
